File: Farmsy/Core/FavoritesStore.swift
import Foundation
import Supabase
import Observation

/// Saved farms — same `favorites(user_id, farm_osm_id)` table the website
/// uses, so hearts stay in sync across web and app.
@MainActor
@Observable
final class FavoritesStore {
    private(set) var osmIds: Set<String> = []

    private struct Row: Decodable {
        let farmOsmId: String
        enum CodingKeys: String, CodingKey { case farmOsmId = "farm_osm_id" }
    }

    func load(userId: UUID) async {
        do {
            let rows: [Row] = try await supabase
                .from("favorites")
                .select("farm_osm_id")
                .eq("user_id", value: userId)
                .execute()
                .value
            osmIds = Set(rows.map(\.farmOsmId))
        } catch {
            // Non-fatal: leave the current set untouched.
        }
    }

    func isSaved(_ osmId: String) -> Bool { osmIds.contains(osmId) }

    func toggle(_ osmId: String, userId: UUID) async {
        if osmIds.contains(osmId) {
            osmIds.remove(osmId)
            do {
                try await supabase
                    .from("favorites")
                    .delete()
                    .eq("user_id", value: userId)
                    .eq("farm_osm_id", value: osmId)
                    .execute()
            } catch {
                osmIds.insert(osmId) // roll back the optimistic update
            }
        } else {
            osmIds.insert(osmId)
            do {
                try await supabase
                    .from("favorites")
                    .insert(["user_id": userId.uuidString, "farm_osm_id": osmId])
                    .execute()
            } catch {
                osmIds.remove(osmId)
            }
        }
    }

    func clear() { osmIds = [] }
}

File: android/app/src/main/java/app/farmsy/android/core/FavoritesStore.kt
package app.farmsy.android.core

import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/// Saved farms — same `favorites(user_id, farm_osm_id)` table the website
/// uses, so hearts stay in sync across web and app. Mirrors FavoritesStore.swift.
class FavoritesStore {

    private val _osmIds = MutableStateFlow<Set<String>>(emptySet())
    val osmIds: StateFlow<Set<String>> = _osmIds.asStateFlow()

    @Serializable
    private data class Row(@SerialName("farm_osm_id") val farmOsmId: String)

    suspend fun load(userId: String) {
        runCatching {
            val rows = supabase.from("favorites")
                .select(columns = Columns.list("farm_osm_id")) {
                    filter { eq("user_id", userId) }
                }
                .decodeList<Row>()
            _osmIds.value = rows.map { it.farmOsmId }.toSet()
        } // Non-fatal: leave the current set untouched.
    }

    fun isSaved(osmId: String): Boolean = _osmIds.value.contains(osmId)

    suspend fun toggle(osmId: String, userId: String) {
        if (_osmIds.value.contains(osmId)) {
            _osmIds.value = _osmIds.value - osmId
            runCatching {
                supabase.from("favorites").delete {
                    filter { eq("user_id", userId); eq("farm_osm_id", osmId) }
                }
            }.onFailure { _osmIds.value = _osmIds.value + osmId } // roll back
        } else {
            _osmIds.value = _osmIds.value + osmId
            runCatching {
                supabase.from("favorites").insert(
                    buildJsonObject {
                        put("user_id", JsonPrimitive(userId))
                        put("farm_osm_id", JsonPrimitive(osmId))
                    }
                )
            }.onFailure { _osmIds.value = _osmIds.value - osmId }
        }
    }

    fun clear() { _osmIds.value = emptySet() }
}
